text/3349-cfg-os-version.md

```diff
@@ -0,0 +1,142 @@
+
+The above example works exactly the same way with the other platform API `cfg` predicates just with different values and different target support.
+
+These predicates do not assume any semantic versioning information. The specified predicates are simply listed in order. The only semantics that are assumed is that code compiled with the `cfg` predicates works for all versions greater than or equal to that version.
+
+**Note:** Here it would be important to link to documentation showing the `cfg` predicates and the different version strings that are supported.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The various target API version `cfg` predicates allow users to conditionally compile code based on the API version supported by the target platform.  Each platform is responsible for defining a default key, a set of keys it supports, and functions that are able to compare the version strings they use.  A set of comparison functions can be provided by `rustc` for common formats such as 2- and 3-part semantic versioning.  When a platform detects a key it doesn’t support it will return `false` and emit a warning.
+
+When a target is being built the actual API versions will be set via the following methods, in decreasing order of precedence:
+* Command line
+* Cargo.toml target sections
+* Target platform defaults
+
+## Versioning Schema
+
+Version strings can take on nearly any form and while there are some standard formats, such as semantic versioning or release dates, projects can change schemas or provide aliases for some or all of their releases.  Because of this diversity in version strings each platform will be responsible for defining a type implementing `FromStr`, `Display`, and `Ord` for each key they support (or using one of the pre-defined types).
+
+## Storing Information in Metadata
+
+The values for the various `cfg` requirements must be stored in a crate’s metadata if the value is set.  If no value is set the platform’s default API versions are used.  If no value is set it will default to the maximum of all the crate’s transitive dependencies’ requirements.
+
+## Future Compatibility
+
+The functions for parsing and comparing version strings will need to be updated whenever a new API is added, when the version format changes, or when new aliases need to be added.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Based on the outcome of various discussions around this RFC it may be necessary to make non-backwards-compatible changes to various parts of `rustc`.  If this occurs older versions of the compiler will not be able to handle crates that make use of these new features.
```

Comment:
I don't quite understand this. I don't think we call using new features as backwards-incompatible. Can you explain what it means? Or, if this is just saying that old compilers can't compile code from the future, then I don't think this needs to be stated. There is no intent to have the compiler be forwards-compatible for language changes.

Reply:
This language is mostly from the original draft of the RFC.  There the proposal included new syntax for annotations (I believe; please correct me if I'm wrong @rylev ) that might have caused backwards-compatibility issues.  I've left this in here to hedge our bets.

If reviewers are confident that we can proceed without needing to make backwards-incompatible changes it'd probably be best to remove this paragraph.

Reply:
This has been removed and replaced with a list of incurred maintenance costs.